Implement the PowerManagingActor

shsms opened the pull request:
This PR add a PowerManagingActor implementation with the Matryoshka
algorithm, that uses actor priorities to resolve conflicting power
requests when multiple actors are trying to use the same batteries.

The Matryoshka algorithm requires actors to specify a desired power
value, and optionally, bounds to limit the range of power values
available for lower priority actors.

Some of the other changes in this PR are:

- updates the `microgrid.battery_pool()` method to accept a priority
  value, and return a new battery pool instance with each call, while
  using a shared reference store class internally, that helps reuse
  formulas, metric aggregators and battery status tracking tasks.

- update BatteryPool's control methods to send requests to the
  PowerManagingActor rather than the PowerDistributingActor.

- Rename the control methods in the BatteryPool from
  `set_power/charge/discharge` -> `propose_{power/charge/discharge}`,
  and update them to accept optional bounds to apply on lower priority
  actors.  The new methods also don't accept a `adjust_power` flag,
  because the `PowerManagingActor` will always adjust power to fit
  within the available bounds.

- Rename `BatteryPool.power_bounds` to a private method called
  `_system_power_bounds`, because it will be used only by the
  PowerManager as the top level bounds.

- Add `BatteryPool.power_status`, which streams
  - bounds adjusted to an actors priority,
  - the latest target power for a set of batteries,
  - the result from the power distributor for the last call.

- Update the PowerDistributingActor to have a single input and output
  channel, because the PowerManagingActor will be the only component
  that uses the PowerDistributingActor from now on.

Closes #161

shsms committed:
Add SortedSet implementation

It uses a Left-Leaning Red Black tree as the underlying algorithm.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Implement PowerManagingActor using the Matryoshka algorithm

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update PowerDistributingActor to have a single output

Because it will be communicating only with the PowerManager.  This
eliminates the need for a namespace field, and for the channel
registry in the PowerDistributingActor.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update `_DataPipeline` to manage `PowerManagingActor` instance

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add PowerDistributor request params to PowerManager proposals

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update BatteryPool to send requests to the PowerManager

PowerManager still forwards results directly from the
PowerDistributor.  This will change in a subsequent commit.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Improve task cleanups in the data pipeline and in test mocks

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove left-over None-typing in BatteryPool's MetricAggregator

It looks like only the `MetricCalculator`s were updated in a previous
PR (#612) and some parts were still pending.  This is taken care of
here.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add some basic tests for the updated BatteryPool control methods

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Track power bounds from BatteryPool for battery sets

These power bounds will be passed to the algorithm in a subsequent
commit.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Mock `_add_bounds_tracker` in the tests for the PowerManagingActor

And use a mock `_system_bounds` value to use with the test.

The tests for the PowerManagingActor are really just testing the
algorithm and don't use the MockMicrogrid, and `_add_bounds_tracker`
will not be able to track system bounds, causing the tests to fail.

Higher level tests using the MockMicrogrid are done as part of the
tests for the control interface of the BatteryPool.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Use system bounds as highest-priority in the Matryoshka implementation

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Test BatteryPool control methods with lower level streaming mocks

This is necessary because, now that the PowerManagingActor requires
system bounds coming from the battery pool,  we need the bounds
calculation part of the battery pool to receive mock raw data as well.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove `adjust_power` as a parameter to BatteryPool's power methods

Also remove `adjust_power` as a field in `Proposal`s to the
`PowerManagingActor`.

This is because the `PowerManagingActor` will always adjust power if
there's an overlap between the requested bounds and the higher
priority bounds, and if the preferred power is outside the intersected
bounds.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove duplicate check for Algorithm in PowerManagingActor

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Rename `BatteryPool.power_bounds` to `_system_power_bounds`

This marks the method as system bounds, and marks it private so that
users will always use the new `power_bounds` methods that would stream
bounds data from the PowerManagingActor.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Report `target_power` and `available_bounds` based on priority

Actors with different priorities will have different available bounds,
based on the limits set by higher priority actors.

This commit adds a `Algorithm.get_status` method that calculates and
returns a report with the target power and available bounds for a
given set of batteries, for a given priority.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Stream status reports corresponding to received requests

This streaming mechanism will be exposed through a method on the
BatteryPool.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Replace `handle_proposal` to `get_target_power`

`get_target_power` returns a target power only if there is a new
target power.  If the target power hasn't changed since the last call,
it returns `None`.

Its `proposal` parameter is optional.  This allows us to call this
function when the system bounds have changed, when no new proposals
have been received, which helps ensure that the target power is always
within the latest bounds.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Rename `_send_report` to `_send_reports`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Make `_add_bounds_tracker` sync, to fix race condition

When there are many simultaneous requests, there was a race condition
that caused multiple bounds trackers to be started for the same set of
batteries.

This commit also fixes some crashes because of missing defaults, when
subscriptions to bounds are made, before any proposals have been
received.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Validate and update target power when system bounds change

Earlier, target power was recalculated only when there was a new
proposal.  But when system bounds change, the state of the system
could become invalid, because the target power is outside the bounds.

This commit fixes the issue by calculating the target power also when
the system bounds change.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Calculate target power only if at least one proposal has been received

When no proposals with preferred power values have been received, the
power manager should not automatically send out zero power.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Use a narrow `ReceiverFetcher` interface to type data stream methods

This allows us to type methods the same even if they different
objects, as long as they all have a `new_receiver` method.

This can't be used to replace the typing of methods that return
`FormulaEngine`s, because they also expose other methods for
composing formulas engines.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Implement a `ReceiverFetcher` that uses an underlying channel registry

This will be used to expose power manager bounds streams through the
battery pool.

This commit also adds a default argument `maxsize` to the channel
registry's `new_receiver` method, forwarded to the channel's
`new_receiver` method.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Expose bounds from PowerManager through the BatteryPool

Also update the battery pool control method tests to use the new
bounds streams.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove `power_distribution_results` from the battery pool

The results from the power distributing actor would instead be used by
the power manager to provide more information through its reporting
stream.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Ignore unknown battery bucket in the absence of system bounds

If a proposal arrives for a previously unknown battery bucket, and
there are no system bounds available for the bucket, then ignore the
proposal.

This commit also improves the code comments for the case where a
previous target power already exists for a given battery bucket,
because in that case, we bring the target power down to zero.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Improve target_power calculation and handling of overlapping bounds

When actors with different priorities have not completely-contained,
but overlapping bounds, use the intersection to find the target_power
closest to the requirements of the lowest priority actor.

This commit also adds some test cases and updates the test to run
directly on a `Matryoshka()` instance, rather than through a
`PowerManagingActor` instance.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Rename `test_power_managing_actor.py` -> `test_matryoshka.py`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Test priority-based bounds calculation for the Matryoshka algorithm

Also fix a bug in bounds calculation that was causing impossible
bounds sent to users.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Refactor target power calculation into a separate method

This simplifies the code, but more importantly, keeps pylint happy.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Rename `available_bounds` to `inclusion_bounds` in PowerManager Reports

They are still the available inclusion limits based on the actor's
priority, but the new (old) name makes it easier to contrast with
exclusion bounds, which will be introduced in the next commits.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Allow `None` values in PowerManager reporting streams

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Replace `Bounds` classes in BatteryPool/PowerManager with a single one

The new `Bounds` class is generic over quantities, and is defined in
the `frequenz.sdk.timeseries` package.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Expose `exclusion_bounds` through PowerManager reporting streams

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Expose power distribution results through PowerManager reporting streams

Failure messages need to trigger a resending of target power to the
PowerDistributor.  This is done in a subsequent commit.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Resend request if PowerDistributor returns PartialFailure

- When there is an issue with any of the batteries, the power
distributor would mark the battery as temporarily broken and update
the battery status stream.

- Changes to the battery status stream would trigger big changes to
the system bounds calculated in the battery pool.

- Changes to the system bounds would not affect the target power
unless it was very big to begin with, so we need to resend the target
power to the power distributor, whenever a partial failure is received
from the power distributor.

- In the next run, the power distributor won't use the broken battery
and will return `Success`.

- If all batteries are broken and the power distributor returns an
`Error` instead, then the bounds will go down to zero, and it will
force a target power to get sent out already.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add a BatteryPoolWrapper to be the external interface of BatteryPool

This would allow for actors interested in operating on the same set of
batteries to share underlying formula engine and metric calculator
instances, but without having to specify their individual priorities
with each request.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Return BatteryPoolWrapper from the data pipeline instead of BatteryPool

This allows users to get a custom object that will keep track of the
user's source_id and priorities, without having to duplicate the
formula engine and metric calculators.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove unused method from BatteryPool

These are now exposed through the BatteryPoolWrapper.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
Turned out to be a lot of work, especially outside the power manager, getting everything connected.  I will set up a call next week to go through the code to make reviewing easier.

A number of parts in the battery pool interface need new names.  We can discuss those during the call as well.

I still need to add some tests, and probably some documentation on how to use priorities, and how the algorithm works, etc, but I think it is ready for review already.

shsms committed:
Accept `None` values in BatteryPool control methods

This allows lower priority actors to completely yield control, which
causes the algorithm to use the perferred power of the higher priority
actor as the target power.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Accept `Bounds` object in `set_power`, instead of separate values

This allows users to specify `None` when they want to set bounds only
in one side, for example, if they only want a minimum charger power of
10kW, and don't care about the maximum.

This commit also makes the `Bounds` type generic without restrictions,
so that it can be specialized for more complex types like
`Bounds[Power | None]`.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Sort `set_power` call order in battery pool tests, before checking

This is because since the m:n battery-inverter ordering was merged,
the PowerDistributingActor is no longer calling
microgridApiClient.set_power in sequential order.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Use `-sys.maxsize - 1` as default priority for BatteryPool

When no priorities are specified, we should use the lowest possible
priority.  Zero was supposed to be this lowest possible value, until
it was mentioned that negative priorities should be possible as well.

This commit also updates the BatteryPool's `charge` and `discharge` to
use the pool's priority rather than the minimum.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

Marenz commented on src/frequenz/sdk/actor/_power_managing/_base_classes.py:
```diff
@@ -0,0 +1,59 @@
+class Proposal:
+    """A proposal for a battery to be charged or discharged."""
+
+    source_id: str
```
Especially the source id should be documented imo

shsms replied:
done in this commit: https://github.com/frequenz-floss/frequenz-sdk-python/pull/692/commits/1e7f54bb7be45f7f4ddb3170cfe16d501530ef3f

shsms committed:
Rename `power_bounds` -> `power_status` and update docs

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

Marenz commented on src/frequenz/sdk/actor/_power_managing/_matryoshka.py:
```diff
@@ -0,0 +1,58 @@
+
+    @override
+    def handle_proposal(self, proposal: Proposal) -> Power:
+        """Handle a proposal.
```
I feel this could also use some more elaboration how it ends up with the target power it returns

shsms replied:
added docs in the module level.

shsms committed:
Rename `set_power/(dis)charge` -> `propose_{power,(dis)charge}`

These are methods in the battery pool, that now go through the power
manager, which tries to reconcile the requests and find a target
power.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

Marenz commented on tests/actor/_power_managing/test_power_managing_actor.py:
```diff
@@ -0,0 +1,56 @@
+        await case(priority=2, power=25.0, bounds=(25.0, 50.0), expected=25.0)
+        await case(priority=1, power=20.0, bounds=(20.0, 50.0), expected=25.0)
+        await case(priority=3, power=10.0, bounds=(10.0, 15.0), expected=10.0)
+        await case(priority=3, power=10.0, bounds=(10.0, 22.0), expected=20.0)
```
Why does this change to 20?

shsms committed:
Accept `name` as parameter in call to `microgrid.battery_pool()`

And explicitly create a new source_id for every new instance of
BatteryPoolWrapper.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

Marenz commented on tests/actor/_power_managing/test_power_managing_actor.py:
```diff
@@ -0,0 +1,56 @@
+        await case(priority=1, power=20.0, bounds=(20.0, 50.0), expected=25.0)
+        await case(priority=3, power=10.0, bounds=(10.0, 15.0), expected=10.0)
+        await case(priority=3, power=10.0, bounds=(10.0, 22.0), expected=20.0)
+        await case(priority=1, power=30.0, bounds=(20.0, 50.0), expected=10.0)
```
and why does this now become 10?

Marenz commented on src/frequenz/sdk/actor/_power_managing/_base_classes.py:
```diff
@@ -18,9 +19,35 @@ class Proposal:
+    fail and be replied to with an `OutOfBound` result.
+    """
+
+    include_broken_batteries: bool = False
```
Didn't we want to remove all broken battery related code?

shsms replied:
Yes, didn't want to hide it here, because then we might not have the motivation to remove it from the PowerDistributor later.  I'll pick that up after this PR.

Marenz commented on src/frequenz/sdk/actor/_power_managing/_matryoshka.py:
```diff
@@ -53,6 +53,7 @@ def handle_proposal(self, proposal: Proposal) -> Power:
             target_power = next_proposal.preferred_power
-            lower_bound = next_proposal.bounds[0]
-            upper_bound = next_proposal.bounds[1]
+            if next_proposal.bounds:
```
Did you sneak in a little null check in this commit? ^^

Marenz commented on src/frequenz/sdk/actor/_power_managing/_power_managing_actor.py:
```diff
@@ -26,35 +28,59 @@
+            channel_registry: The channel registry.
             algorithm: The power management algorithm to use.
 
+
```
superfluous blank line

shsms replied:
removed

Marenz commented on src/frequenz/sdk/actor/_power_managing/_power_managing_actor.py:
```diff
@@ -26,35 +28,59 @@
         self._bound_tracker_tasks: dict[frozenset[int], asyncio.Task[None]] = {}
+        self._subscriptions: dict[frozenset[int], dict[int, Sender[Report]]] = {}
+
+        self._algorithm: BaseAlgorithm = Matryoshka()
```
You removed this in an earlier commit, only to add it back here? Oo

Marenz commented on src/frequenz/sdk/actor/_power_managing/_power_managing_actor.py:
```diff
@@ -0,0 +1,221 @@
+        # pylint: enable=protected-access
+
+        # Fetch the latest system bounds once, before starting the bounds tracker task,
+        # so that when this function returns, there's already some bounds available.
```
This comment is no longer correct

shsms replied:
true, removed

Marenz commented on src/frequenz/sdk/actor/_power_managing/_power_managing_actor.py:
```diff
@@ -96,7 +97,7 @@ async def _bounds_tracker(
             await self._send_reports(battery_ids)
 
-    async def _add_bounds_tracker(self, battery_ids: frozenset[int]) -> None:
+    def _add_bounds_tracker(self, battery_ids: frozenset[int]) -> None:
```
I _was_ wondering if the `async` here might create limits/problems later

Marenz commented on src/frequenz/sdk/timeseries/battery_pool/battery_pool.py:
```diff
@@ -447,6 +454,28 @@ def capacity(self) -> ReceiverFetcher[Sample[Energy]]:
+        """Get a receiver to receive new power bounds when they change.
+
+        These bounds are the bounds specified the power manager for actors with the
+        given priority.
```
Can you add more words to this sentence so it doesn't read like I am hiking barefoot through a field of sharp stones?

shsms replied:
:D wow, that's so funny, the stuff I've written.  But already fixed.  And doing the other doc stuff as well.

Marenz commented on src/frequenz/sdk/actor/_power_managing/_matryoshka.py:
```diff
@@ -49,6 +52,21 @@ def get_target_power(
+                        "ignored.",
+                        battery_ids,
+                    )
+                return None
```
Could it happen that someone sets power before the system bounds are initialized? Is there a way to make sure they are initialized before using?

shsms replied:
yes, I think the trick is for actors to get the bounds first, ensure their requests comply with the bounds, then make the proposal.

but even if people don't do that,  only the first request will get ignored (and logged) because there are no bounds.  because as soon as a proposal comes in, we start tracking bounds for that set of batteries.

Marenz commented on src/frequenz/sdk/actor/_power_managing/_base_classes.py:
```diff
@@ -138,6 +139,8 @@ def get_target_power(
             proposal: If given, the proposal to added to the bucket, before the target
                 power is calculated.
             system_bounds: The system bounds for the batteries in the proposal.
+            must_return_power: If `True`, the algorithm must return a target power,
```
Is there any reason not to always do this?

shsms replied:
This was supposed to reduce the number of requests going out to the power distributor, especially when there are multiple actors repeating their requests.  

But that just reminded me, we can't have this yet,  because requests to the microgrid service will expire, and we have to keep repeating them periodically.  There's a ticket for this #293, and once we have that,  we can enable `must_return_power` back.  For now,  I've added a commit https://github.com/frequenz-floss/frequenz-sdk-python/pull/692/commits/a8bf82d59f7869dea517257acddcc634f59054c9 that makes `must_return_power` always true.

Marenz commented on src/frequenz/sdk/timeseries/battery_pool/_battery_pool_wrapper.py:
```diff
@@ -57,52 +57,65 @@ def __init__(self, battery_pool: BatteryPool, source_id: str | None, priority: i
+        The preferred power of lower priority actors will take precedence as long as
+        they respect the bounds set by higher priority actors.  If lower priority actors
+        request power values outside of the bounds set by higher priority actors, the
+        target power will be closest value to the preferred power that is within the
```
will be *the* closest value

shsms replied:
fixed

Marenz commented on src/frequenz/sdk/timeseries/battery_pool/_battery_pool_wrapper.py:
```diff
@@ -44,17 +44,17 @@ class BatteryPoolWrapper:
         Args:
             battery_pool: The battery pool to wrap.
-            source_id: The source ID to use for the requests made by this wrapper.
+            name: The source ID to use for the requests made by this wrapper.
```
It's not yet the source id though, only after the unique_id is added, right? Needs docs update :)

shsms replied:
updated

Marenz commented on src/frequenz/sdk/actor/_power_managing/_matryoshka.py:
```diff
@@ -0,0 +1,200 @@
+            NotImplementedError: When the proposal contains battery IDs that are
+                already part of another bucket.
+        """
+        if battery_ids not in self._battery_buckets:
```
Could we move the checks in an extra function to make this one easier to read?

shsms replied:
done

Marenz reviewed (commented):
I had a few comments, overall nice work

shsms committed:
Improve documentation in multiple parts

This includes a new admonition for `microgrid.battery_pool()` that
higher values mean higher priorities.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Rename `BatteryPool` -> `BatteryPoolReferenceStore`

Also make it a private module.  In a subsequent commit, the current
`BatteryPoolWrapper` will be renamed to `BatteryPool`, as the only
public interface.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Rename `BatteryPoolWrapper` -> `BatteryPool`

Also improve its public facing documentation.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Re-export `ReceiverFetcher` to make it visible in documentation

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Send requests to PowerDistributor even if target power hasn't changed

This is required until
https://github.com/frequenz-floss/frequenz-sdk-python/issues/293 is
done.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Refactor battery_id validation to improve readability

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update RELEASE_NOTES.md

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

llucax commented on RELEASE_NOTES.md:
```diff
@@ -31,12 +45,11 @@
   batteries.
   This means that one or more inverters can be connected to one or more batteries.
 
+- A PowerManagingActor implementation
```
Optional:

```suggestion
- A `PowerManagingActor` implementation
```

shsms replied:
Done in: https://github.com/frequenz-floss/frequenz-sdk-python/pull/692/commits/8d6edef480e87c12a2523f9ae009fd2bd7f0b5eb

llucax commented on benchmarks/power_distribution/power_distributor.py:
```diff
@@ -47,10 +51,12 @@ async def send_requests(batteries: set[int], request_num: int) -> list[Result]:
     for _ in range(request_num):
-        await battery_pool.set_power(Power(float(random.randrange(100000, 1000000))))
+        await battery_pool.propose_power(
+            Power(float(random.randrange(100000, 1000000)))
```
Optional (and specially because this PR is not really adding these numbers): this will soon fail when we incorporate [`flake8-numbers`](https://github.com/frequenz-floss/flake8-numbers) (and I think is a perfect example of where it adds clarity :D):
```suggestion
            Power(float(random.randrange(100_000, 1_000_000)))
```

shsms replied:
It is not new, can be fixed in flake8 PR.

llucax commented on src/frequenz/sdk/_internal/_channels.py:
```diff
@@ -0,0 +1,26 @@
+T = typing.TypeVar("T")
+
+
+class ReceiverFetcher(typing.Generic[T], typing.Protocol):
+    """An interface that just exposes a `new_receiver` method."""
+
+    @abc.abstractmethod
+    def new_receiver(self, maxsize: int = 50) -> Receiver[T]:
+        """Get a receiver from the channel.
+
+        Args:
+            maxsize: The maximum size of the receiver.
+
+        Returns:
+            A receiver instance.
+        """
```
`Protocol`s are very nice but I think we should be careful in how we introduce/use them. `Protocol`s are particularly useful to keeps type very close with who use them (and to reduce the required interface to the minimum) and to define inputs more than outputs.

This is because you don't need to inherit from the protocol for a class to be considered an *instance* of a protocol, so you can make 2 different libraries interact much more easily, For example, if `liba` has a function `liba.make_bark(dog: liba.Dog) -> None: dog.bark()` and `libb` has a function `libb.parse_dog(dog: str) -> libb.Dog`, then without protocols you can't do: `liba.make_bark(libb.parse_dog())` even if `Dog.bark()` exist in both `liba.Dog` and `libb.Dog`, and if you don't have control over `liba` to make `liba.Dog` inherit from `libb.Dog`, then you need to convert between both. If `liba.Dog` is a protocol instead, then everything works out of the box.

Looking on how you use this, specially since you inherit from it (which it has extra advantages, as your subclass will be checked to have all the necessary methods when the class is parsed instead of used), I don't see any point in making it a `Protocol`. Not sure if you purposely made it a `Protocol` because you anticipate to have interactions like the ones I mentioned above... If not, I mildly suggest to not make this inherit from `Protocol`, as there is one more thing people have to know well how it works and why it is there.

EDIT: I see now that `MetricAggregator` is behaving as a `ReceiverFetcher` but not inheriting from it, now I wonder why you decided to make some inherit from it and some not.

shsms replied:
It was made to provide a single type around `MetricAggregator` and the `ChannelRegistry`.  But you're right, the internal channel registry type doesn't have to derive from `ReceiverFetcher`.  Done here: https://github.com/frequenz-floss/frequenz-sdk-python/pull/692/commits/50fc8d34062b64df7bbb9b6e39f6a983fb7e5dfa

llucax replied:
OK, my point was that probably in this case it made more sense to just not use a `Protocol` and derive both instead, but I guess that's also more likely to end with circular dependencies (I guess that's another nice feature of `Protocol`). I'm still not convinced which is better to be honest, but since I'm not convinced I will leave the topic here.

Almost. If we start using `Protocol` more, actually `ReceiverFetcher` (maybe with another name) could be moved to `channels`, as most channels are also `ReceiverFetcher`s. Also I think `Sender` and `Receiver` could be protocols.

shsms replied:
I agree about moving it to the channels repo, already have a todo item for it,  because there's really no nice place to put it in the sdk, especially because it has to be publicly exposed, and it is a bit weird now, living in `_internal`, but still exposed publicly.

llucax commented on src/frequenz/sdk/actor/_power_managing/_base_classes.py:
```diff
@@ -0,0 +1,189 @@
+    from .. import power_distributing
+
+
+@dataclasses.dataclass(frozen=True)
```
When most attributes of a `dataclass` are not classes but built-in types, I think it might make things more clear if we also use `kw_only=True` to force users use keywords.

```py
# cryptic
ReportRequest("lalala", [1,2,3], 5)
# obvious
ReportRequest(source_id="lalala", battery_ids=[1,2,3], priority=5)
# unless this is always expected to be used like
ReportRequest(get_id(), battery_ids, priority)
# (but well, is hard to ensure in any possible future this holds true)
```

shsms replied:
Done here: https://github.com/frequenz-floss/frequenz-sdk-python/pull/692/commits/6dfde504d93558de8e421c9856c0fdc42ab6d1d7

llucax commented on src/frequenz/sdk/actor/_power_managing/_base_classes.py:
```diff
@@ -0,0 +1,189 @@
+
+
+@dataclasses.dataclass(frozen=True)
+class Report:
```
Here, for example, it is mostly fine, because we'll always at least have:

```py
Report(Power.from_kw(10), Bounds(...), Bounds(...), Result(...))
```
But yeah, with the bounds it is inclusion or exclusion first? We can use `filed` to make only those 2 keyword-only. Then users will need to write:

```py
Report(Power.from_kw(10), Result(...), inclusion_bounds=Bounds(...), exclusion_bounds=Bounds(...))
```

Since they are kw-only now, we could even remove the `_bounds` suffix as it's a bit redundant now (`inclusion=Bounds(...), exclusion=Bounds(...)`), but I don't know, it reads a bit weird without them too.

shsms replied:
Done here: https://github.com/frequenz-floss/frequenz-sdk-python/pull/692/commits/6dfde504d93558de8e421c9856c0fdc42ab6d1d7

llucax commented on src/frequenz/sdk/actor/_power_managing/_base_classes.py:
```diff
@@ -0,0 +1,189 @@
+    target_power: Power | None
+    """The currently set power for the batteries."""
+
+    inclusion_bounds: timeseries.Bounds[Power] | None
+    """The available inclusion bounds for the batteries, for the actor's priority.
+
+    These bounds are adjusted to any restrictions placed by actors with higher
+    priorities.
+    """
+
+    exclusion_bounds: timeseries.Bounds[Power] | None
```
To make bounds keyword-only (**and we need to move them to the end**):
```suggestion
    inclusion_bounds: timeseries.Bounds[Power] | None = dataclasses.field(kw_only=True)
    """The available inclusion bounds for the batteries, for the actor's priority.

    These bounds are adjusted to any restrictions placed by actors with higher
    priorities.
    """

    exclusion_bounds: timeseries.Bounds[Power] | None = dataclasses.field(kw_only=True)
```

shsms replied:
Done here: https://github.com/frequenz-floss/frequenz-sdk-python/pull/692/commits/6dfde504d93558de8e421c9856c0fdc42ab6d1d7

Marenz commented on src/frequenz/sdk/timeseries/battery_pool/_battery_pool_reference_store.py:
```diff
@@ -21,8 +21,15 @@
+    - the formula engine pool and metric calculators.
+    - the tasks for updating the battery status for the metric calculators.
+
+    These are independent of the priority of the actors, and can be shared between
```
The `,` is [wrong](https://www.grammarly.com/blog/comma-before-and/) as the second part of the conjunction is not an independent clause.

shsms replied:
done in: https://github.com/frequenz-floss/frequenz-sdk-python/pull/692/commits/8d6edef480e87c12a2523f9ae009fd2bd7f0b5eb

llucax commented on src/frequenz/sdk/actor/_power_managing/_base_classes.py:
```diff
@@ -0,0 +1,189 @@
+        )
+
+
+class Algorithm(enum.Enum):
```
Not sure if you explained this somewhere else, but otherwise I would mention in a code comment (not class documentation) that we use an enum because at some point this algorithm will need to be serialized.

shsms replied:
Added a couple of lines near the actor parameter here: https://github.com/frequenz-floss/frequenz-sdk-python/pull/692/commits/f268d16e0a1fc901a8e61a7552f95799cf8b4f44

llucax commented on src/frequenz/sdk/actor/_power_managing/_power_managing_actor.py:
```diff
@@ -0,0 +1,230 @@
+
+"""The power manager."""
+
+from __future__ import annotations
```
Is this really needed? I think in Python 3.11+ it is only necessary to use forward references and I don't see any, but maybe I am missing something.

shsms replied:
yes, without it, we have to quote the first occurrence of a type hint that was added with `if TYPE_CHECKING`.

llucax commented on src/frequenz/sdk/microgrid/_data_pipeline.py:
```diff
@@ -173,42 +189,93 @@ def ev_charger_pool(
         Args:
             battery_ids: Optional set of IDs of batteries to be managed by the
                 BatteryPool.
+            name: An optional name used to identify this instance of the pool or a
+                corresponding actor in the logs.
```
OK, so you finally went with always creating a UUID and just adding the name to it if it was provided, right? So this means there is no way to call `battery_pool()` 2 times as the same "user" (i.e. the request being considered to come from the same source).

This means to do that we need to save the `battery_pool` object we just *created* and we don't support the case where someone wants to be considered the same source in the eyes of the power manager from different nodes/controllers.

Just checking if we are all in the same page and fine with going this route...

shsms replied:
Yes,  but this reminds me, we need to add logging in the actor to warn users when there are multiple proposals with the same priority, but with different `source_id`.  Made an issue, so this PR doesn't keep dragging: https://github.com/frequenz-floss/frequenz-sdk-python/issues/714

Marenz commented on src/frequenz/sdk/microgrid/_data_pipeline.py:
```diff
@@ -191,13 +191,11 @@ def battery_pool(
-        and returned.
-
-        The BatteryPool is wrapped in a new `BatteryPoolWrapper` instance each time.
+        If a BatteryPoolReferenceStore instance for the given battery ids doesn't exist,
```
back-tick-quotes around `BatteryPoolReferenceStore` would be nice.

Marenz replied:
and `BatteryPool` I suppose :)

shsms replied:
Done in: https://github.com/frequenz-floss/frequenz-sdk-python/pull/692/commits/8d6edef480e87c12a2523f9ae009fd2bd7f0b5eb

llucax commented on src/frequenz/sdk/microgrid/_data_pipeline.py:
```diff
@@ -173,42 +189,93 @@ def ev_charger_pool(
         self,
         battery_ids: abc.Set[int] | None = None,
+        name: str | None = None,
+        priority: int = -sys.maxsize - 1,
```
Any reason not to make the `priority` a `float`, so we don't end up writing *BASIC* to make sure we can sneak in some new priority between 2 existing priorities (priority 10, 20, 30, 40.... :D).

Also I don´t get the default. Why not 0?

shsms replied:
Don't know, with all the float comparison issues.  We can just recommend 6 digit priorities, so there's always enough space between values.

llucax replied:
Considering that:

* We only care about ordering
* If 2 priorities are conceptually equal, we don't care which one is considered higher than the other (in cases where `float` could not compare equal for 2 equal numbers, like `assert 0.1 + 0.2 > 0.3`)
* We'll probably be using small (but not too small) literals
 
It should be reasonably safe to use `float`s IMHO.

The problem with integers is if we end up with 2 actors using consecutive numbers there is no (external) way to sneak something between both, and you might not have control over those 2 actors to go change the code and redeploy. I we would want to be extra safe, we could add a `Priority(float)` object that checks if the `float` is perfectly representable and raises if it's not, but as a first step I would use `float`.

llucax replied:
BTW, I'm fine with merging as is and discuss this separate and do a follow-up PR if we decide to change it.

llucax replied:
Also, I just saw the commit that makes `-sys.maxsize - 1` the default. I'm not sure why we want to have the lowest possible priority ever the default, that is not explained in the commit message.

shsms replied:
**int vs float:**

My feeling has always been that priorities have to be externally controlled, possibly through UI config.  Because actor implementations by themselves can't know what their priorities are.

But going from int to float is not a breaking change, so I agree that we can discuss separately.

**default priority:**

This is of course subjective, and is based on some further opinions: 
- The default priority is used only when an actor knows it will not be sharing batteries with other actors.  
- Most actors don't know this, so they should provide a way to configure their priorities.  
- If an actor uses the default priority (i.e doesn't care about its priority), it should be lower than any other actor, even the ones with negative priorities, because negative priorities doesn't mean that they're low priority, just that the priority numbers are negative.  
- If actors care about priorities, they should provide a way to specify one, else they get put in the bottom of the pile.

llucax replied:
Yeah, let's leave it for when we have a better idea of how this will be used and who will be setting priorities and how.

llucax commented on src/frequenz/sdk/timeseries/battery_pool/_battery_pool.py:
```diff
@@ -36,143 +33,132 @@
-            adjust_power: If True, the power will be adjusted to fit the power bounds,
-                if necessary. If False, then power requests outside the bounds will be
-                rejected.
+            power: The power to propose for the batteries in the pool.  If None, the
+                proposed power of higher priority actors will take precedence as the
+                target power.
```
I find the comment about `None` a bit confusing. Isn't the intent of passing `None` to sort of *withdraw* a proposal, like not participating in the decision of how much power to set? If so I would mention something more explicit along these lines to make more clear what's the use case, not only what is the effect.

Also nitpick:
```suggestion
            power: The power to propose for the batteries in the pool.  If `None`, the
                proposed power of higher priority actors will take precedence as the
                target power.
```

shsms replied:
It is a bit tricky, because there could be bounds even if power is `None`.  If both of them are `None`, that's like withdrawing the proposal,  else it is just "not having a perferred power".

llucax replied:
OK, I see, to be honest I didn't read all the docs, so if this is well explained somewhere, we should be fine. It is complicated so we can't have a full explanation everywhere, agreed.

shsms replied:
maybe this looks better?

```suggestion
            power: The power to propose for the batteries in the pool.  If `None`, this
                proposal will not have any effect on the target power, unless bounds
                are specified.  If both are `None`, it is equivalent to not having a 
                proposal or withdrawing a previous one.
```

llucax replied:
Yup!

shsms replied:
done in: https://github.com/frequenz-floss/frequenz-sdk-python/commit/79ecc0d2

llucax commented on src/frequenz/sdk/timeseries/battery_pool/_battery_pool.py:
```diff
@@ -36,143 +33,132 @@
+        system power bounds for the batteries.
+
+        The result of the request can be accessed using the receiver returned from the
+        `power_status` method, which also streams the bounds that an actor should
```
In general, I think it would be good to start including more cross-referencing in new code:
```suggestion
        [`power_status`][frequenz.sdk.timeseries.battery_pool.power_status] method, which also streams the bounds that an actor should
```

shsms replied:
Done here: 9eabfb59e7511adf599a704bfcf4646467c46527

llucax commented on src/frequenz/sdk/timeseries/battery_pool/_battery_pool_reference_store.py:
```diff
@@ -0,0 +1,138 @@
+
+    This includes ownership of
+    - the formula engine pool and metric calculators.
+    - the tasks for updating the battery status for the metric calculators.
```
Oh, this has `BackgroundService` written all over it.

shsms replied:
made an issue: https://github.com/frequenz-floss/frequenz-sdk-python/issues/715

llucax commented on src/frequenz/sdk/timeseries/battery_pool/_battery_pool_reference_store.py:
```diff
@@ -0,0 +1,138 @@
+            resampler_subscription_sender,
+        )
+
+    async def stop(self) -> None:
+        """Stop all pending async tasks."""
+        tasks_to_stop: list[Awaitable[Any]] = [
+            method.stop() for method in self._active_methods.values()
+        ]
+        tasks_to_stop.append(self._formula_pool.stop())
+        if self._update_battery_status_task:
+            tasks_to_stop.append(cancel_and_await(self._update_battery_status_task))
+        await asyncio.gather(*tasks_to_stop)
```
And this.

I guess at this point it would delay the PR a bit, so I wouldn't block it because of this, but it would be good to have a follow-up PR converting this into a `BackgroundService`. The whole point of introducing that class was to unify how we build classes like this.

shsms replied:
made an issue: https://github.com/frequenz-floss/frequenz-sdk-python/issues/715

llucax commented on tests/actor/_power_managing/_sorted_set.py:
```diff
@@ -0,0 +1,39 @@
+from frequenz.sdk.actor._power_managing import _sorted_set
+
+
+def test_sorted_set() -> None:
```
I guess this is a pretty good candidate for *hypothesis* (many inputs that could be tested and just a few simple properties to test for).

shsms replied:
made an issue: https://github.com/frequenz-floss/frequenz-sdk-python/issues/716

llucax commented on tests/actor/_power_managing/test_matryoshka.py:
```diff
@@ -0,0 +1,162 @@
+    )
+    test_bounds(priority=1, expected_power=25.0, expected_bounds=(25.0, 50.0))
+
+    test_tgt_power(priority=3, power=10.0, bounds=(10.0, 15.0), expected=15.0)
```
Ha! This is confusing, I think it makes sense but it makes you think about priorities a bit differently from what one might expect. So high priority means high priority for bounds but low priority for target power, right? So this is a higher priority request so the bounds are narrowed but the actual target power is the closest to the lowest priority request instead, right? And this is why this high prio request gets `15` instead of `10`, which is what it actually wanted?

shsms replied:
Right.  If that's not the case, and we just prioritize preferred powers of higher pri actors, then lower pri actors might as well not bother sending any request, and we don't need bounds,  and probably no power manager, and actually the lower pri actors don't need to run at all. :D

llucax reviewed (commented):
OK, I finished a quick look. I guess all is sort of optional, at least nothing major but a few things I find important:

* Keyword-only stuff (only if there is any expectation to be used by end users, which I think there aren't, so it should be fine as it is)
* `priority` as `float` and default
* Make sure we are aligned about the `source_id`/`name`
* Using `BackgroundService` for the classes that it makes sense (OK to do in a separate PR, as lower prio as we still have a couple of classes that are not converted yet, but from now on we should always use `BackgroundService` from the start for stuff handling tasks, having a  `stop()`, etc.)

I think that's it, the rest is more cosmetic or not that important.

shsms committed:
Backquote keywords in documentation

Also fix a documentation grammar mistake involving the removal of a
comma.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Derive `_RegistryReceiverFetcher` from `Generic[T]`

Because `ReceiverFetcher` is a `Protocol` type,
`_RegistryReceiverFetcher` will fit as a `ReceiverFetcher` without
having to derive from it.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Make data-class constructors kw-only

For these classes: `ReportRequest`, `Report`, `Proposal`.

Also update the uses of these classes that weren't already using kw
args.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add code comment explaining the use of an Enum for Algorithm

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add more cross-referencing in BatteryPool docstrings

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update `BatteryPool.propose_power` docstring to be less ambiguous

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

Marenz commented on tests/timeseries/_battery_pool/test_battery_pool_control_methods.py:
```diff
@@ -0,0 +1,370 @@
+        # not used.
+        #
+        # It will be replaced by a reporting streaming from the PowerManager in a
+        # subsequent commit.
```
Is this something you meant to update in this PR?

shsms replied:
oh yes, that whole comment is no longer relevant.  Will remove.

shsms replied:
done

matthias-wende-frequenz commented on src/frequenz/sdk/actor/_power_managing/_base_classes.py:
```diff
@@ -0,0 +1,189 @@
+
+    These bounds will apply to actors with a lower priority, and can be overridden by
+    bounds from actors with a higher priority.  If None, the power bounds will be set to
+    the maximum power of the batteries in the pool.  This is currently and experimental
```
```suggestion
    the maximum power of the batteries in the pool.  This is currently an experimental
```

shsms replied:
I wonder if `bounds` is an experimental feature.  Maybe we should rename it back to `_bounds` in the interface.  But the interface has a new name `propose_power` won't make sense without the bounds.  Is tricky.

matthias-wende-frequenz replied:
at least you should replace the and with an an ;)

shsms replied:
done

matthias-wende-frequenz commented on src/frequenz/sdk/actor/_power_managing/_base_classes.py:
```diff
@@ -0,0 +1,189 @@
+    """The base class for algorithms."""
+
+    @abc.abstractmethod
+    def get_target_power(
```
I wonder why this is called `get_target_power` although it's used to propose proposals. What about `calculate_target_power` or `determine_target_power`?

shsms replied:
True, `calculate_target_power` sounds way better, will rename.

Marenz replied:
ANOTHER rename

Marenz replied:
I am still voting for `request_..`

matthias-wende-frequenz replied:
It's just returning a calculated power. If I'd request something I'd suppose something happens, e.g. updating a db or switching a state .... But in this case nothing happens.
Therefore I'd find request misleading.

shsms replied:
renamed to `calculate_target_power`.  Regarding the renaming from `propose_*` to `request_*` (assuming this is what you're talking about @Marenz),  I'd like it to not block this PR, but I'll start an internal discussion where we can finalize the name and make a separate PR if necessary.

shsms committed:
Rename `get_target_power` to `calculate_target_power`

Also fix a typo and remove an outdated code comment.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Merge branch 'v1.x.x' into power-manager
